File: shell/shell.cpp
#include "shell.h"


void Loop()
{
    std::string line;
    std::vector<std::string> args;
    int status;

    do
    {
        line = ReadLine();
        args = GetArgs(line);
        status = Execute(args);
    } while (status);
}

int SystemCommands(std::vector<std::string> args_orig)
{
        int status;
        pid_t pid = fork(); // cria o processo filho e armazena o PID retornado da função fork

        char **args = new char *[args_orig.size() + 1];
        
        for(int i = 0; i < args_orig.size(); i++)
        {
            args[i] = new char[args_orig[i].size()];
            strcpy(args[i], args_orig[i].c_str());
        }

        args[args_orig.size()] = NULL;

        if (pid < 0) // Falha no fork
        {
            perror("lsh");
            return -1;
        }

        else{
            if (pid != 0) // Caso fork() retorne o PID do filho, ele é o processo pai
            {
                while (true)
                {
                    waitpid(pid, &status, WUNTRACED); // espera o filho terminar o processo
                    if ((WIFSIGNALED(status) || WIFEXITED(status)))
                        break;
                }
            }
            else
            {
                if (execvp(args[0], args) == -1) { // filho executa o comando
                        perror("lsh");
                    }
                    exit(EXIT_FAILURE);

                    /*modifica a imagem do processo filho para o arquivo chamado*/
            }

            for(int i = 0; i < args_orig.size(); i++)
            {
                delete[] args[i];
                args[i] = nullptr;
            }
            
            delete[] args;
            args = nullptr;
        
        return 1;
        }
}

std::string ReadLine()
{

    std::cout <<  "\x1b[35m" <<get_current_dir_name()
              << "> " << "\x1b[0m";
    char *Temp;
    size_t Size = 0;

    if (getline(&Temp, &Size, stdin) == -1)
    {
        return "";
    }

    return Temp;
}

std::vector<std::string> GetArgs(const std::string &_line)
{
    std::vector<std::string> Temp;
    char *TempChar = new char[64];
    strcpy(TempChar, _line.c_str());
    // copia o valor da linha para o ponteiro tempchar
    // a função strtok recebe um ponteiro de char, entao copiamos o conteudo da linha para um ponteiro de char
    char *TempTokens = strtok(TempChar, " \n\b\t\r"); // a primeira chamada da função passa um ponteiro da string e retorna tudo que encontrar antes dos tokens

    while (TempTokens != nullptr)
    {
        Temp.push_back(TempTokens);
        TempTokens = strtok(0, " \n\b\t\r"); // nessa segunda chamada, o argumento 0 é usado para o ponteiro começar de onde parou
    }

    delete[] TempChar;
    delete[] TempTokens;
    // Boa pratica
    TempChar = nullptr;
    TempTokens = nullptr;

    // Temp.push_back((char*)0x00);

    return Temp;
}

int Execute(const std::vector<std::string> &_args)
{
    if (_args.size() == 0)
    {
        return 1;
    }

    if (Commands.find(_args[0]) != Commands.end())
    {
        return Commands[_args[0]](_args);
    }
    
    return SystemCommands(_args);
}

File: shell in python/shell.py
import os
import sys
import datetime

class Shell:

    def ChangeDir(args):
        if len(args) > 1:
            try:
                os.chdir(args[1])
            except Exception as e:
                print('chdir() error: {0}'.format(e))
        return 1

    def Clock(args):
        now = datetime.datetime.now()
        print(now.strftime("%A, %d. %B %Y %I:%M%p"))
        return 1

    def Help(args):
        print('The following are built in: \n')
        for key in Shell.MyCommands.keys():
            print(key)
        return 1

    def Quit(args):
        return 0
    
    MyCommands = {
        'cd' : ChangeDir,
        'quit' : Quit,
        'clock' : Clock,
        'help' : Help
    }

    def __init__(self):
        status = 1
        while(status):
            line = self.ReadLine()
            args = self.GetArgs(line)
            status = self.Execute(args)

    def ReadLine(self):
        return input('\033[33m' + os.getcwd() + '> \033[0m')

    def GetArgs(self, command):
        return command.split(' ')

    def Execute(self, args):
        if(len(args) == 0):
            return 1
        if args[0] in Shell.MyCommands.keys():
            return Shell.MyCommands[args[0]](args)

        return self.SystemCommands(args)

    def SystemCommands(self, args):
        try:
            pid = os.fork()
        except Exception as e:
            print('fork() error : {0}'.format(e))
        
        if pid != 0:
            while(True):
                _, status = os.waitpid(pid, os.WUNTRACED)
                if ((os.WIFSIGNALED(status) or os.WIFEXITED(status))):
                    break
        else:
            try:
                os.execvp(args[0], args)
            except Exception as e:
                print('execvp() error : {0}'.format(e))
            os._exit(-1)

        return 1 
    
shell = Shell()
